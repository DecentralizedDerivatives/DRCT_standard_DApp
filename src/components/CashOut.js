import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Collapse } from 'reactstrap';
import CashOutFormContainer from './CashOutFormContainer';
import { getUserBalance } from '../actions/userActions';

// Use named export for unconnected component for testing
export class CashOut extends Component {
  constructor() {
    super();
    this.state = {
      collapse: false,
      resultsMessage: ''
    };

    this.toggleFormVisibility = this.toggleFormVisibility.bind(this);
  }

  async componentWillMount() {
    await this.props.getUserBalance();
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.cashOutError !== 'null') {
      this.setState({
        resultsMessage: `Error: ${nextProps.cashOutError}`,
        collapse: false
      });
    } else if (nextProps.cashOutTx) {
      this.setState({
        resultsMessage: `Tx receipt: ${nextProps.cashOutTx}`,
        collapse: false
      });
    }
  }

  // Toggle form visibility on button click
  toggleFormVisibility() {
    this.setState({
      resultsMessage: '',
      collapse: !this.state.collapse
    });
  }

  render() {
    return (
      <div className="container">
        <h3 className="user-balance">Your Balance: {this.props.userBalance}</h3>
        <div id="cashout-button">
          <button
            className="btn create-contract-btn"
            onClick={this.toggleFormVisibility}
          >
            Cash Out
          </button>
        </div>
        <Collapse isOpen={this.state.collapse}>
          <div id="cashout-form">
            <h4 className="center-text">Cash Out Request</h4>
            <CashOutFormContainer />
          </div>
        </Collapse>

        {this.state.resultsMessage && (
          <div id="results-message" className="text-center">
            {this.state.resultsMessage}
          </div>
        )}
      </div>
    );
  }
}

CashOut.propTypes = {
  getUserBalance: PropTypes.func.isRequired,
  userBalance: PropTypes.number.isRequired,
  cashOutTx: PropTypes.string,
  cashOutError: PropTypes.string
};

const mapStateToProps = state => ({
  userBalance: state.user.userBalance,
  cashOutTx: state.user.cashOutTx,
  cashOutError: state.user.cashOutError
});

export default connect(
  mapStateToProps,
  { getUserBalance }
)(CashOut);
