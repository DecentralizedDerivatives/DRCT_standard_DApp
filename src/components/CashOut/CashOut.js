import React, {Component} from 'react';
import PropTypes from 'prop-types';
import withStyles from 'material-ui/styles/withStyles';
import Button from 'material-ui/Button';
import Typography from 'material-ui/Typography';
import Dialog, {DialogContent} from 'material-ui/Dialog';
import styles from './styles';
import {Wrapped, web3} from '../../ethereum';

class CashOut extends Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    open: PropTypes.bool.isRequired,
    toggle: PropTypes.func.isRequired,
  };

  static durations = ['One weeks', 'Two weeks'];
  static currency = ['BTC/USD', 'ETH/USD'];

  state = {
    open: false,
    duration: '',
    currency: '',
    amount: 0.1,
    selectedDate: new Date(),
    loading: false,
    disabled: false,
    created: false,
    myBalance:"0"
  };

  handleChange = event => {
    this.setState({[event.target.name]: event.target.value});
  };

  handleDateChange = date => {
    this.setState({selectedDate: date});
  };

  handleTextfieldChange = name => event => {
    this.setState({
      [name]: event.target.value,
    });
  };

  componentWillMount() {
    this.getMyBalance().then((result)=>{
      console.log('res',result);
    });
  }

  getMyBalance= async () =>{
    const wrapped = await Wrapped.deployed();
    const accounts = await web3.eth.getAccounts();
    var _res = await wrapped.balanceOf(accounts[0]);
    return _res.c[0];
  }

  cashOut= async () => {
    const wrapped = await Wrapped.deployed();
    const accounts = await web3.eth.getAccounts();
    let response, error;
    try {
      await wrapped.withdraw(this.state.myBalance,{
        from: accounts[0],
        gas: 4000000
      });
    } catch (err) {
      error = err;
    }
    if (error) {
      console.log(error);
    }
  };

  render() {
    const {classes} = this.props;

    return (
      <div>
        <Dialog
          open={this.props.open}
          onClose={this.props.toggle}
          PaperProps={{className: classes.paper}}
        >
          <DialogContent className={classes.dialogContent}>
            <div className={classes.inputContainer}>
              <Typography className={classes.title}>Cash Out</Typography>
            </div>

            <div className={classes.inputContainer}>
              <Typography className={classes.title}>Amount to withdraw: {this.state.myBalance}</Typography>
            </div>


            <Button
              className={classes.button}
              onClick={this.cashOut}
            >
              <Typography className={classes.buttonText}>Submit</Typography>
            </Button>
            </DialogContent>
        </Dialog>
      </div>
    );
  }
}

export default withStyles(styles)(CashOut);
