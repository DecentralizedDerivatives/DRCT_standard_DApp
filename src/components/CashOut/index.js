import CashOut from './CashOut';

export default CashOut;
