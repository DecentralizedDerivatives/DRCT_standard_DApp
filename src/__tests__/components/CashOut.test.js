// TODO: check
import CashOut from '../../components/CashOut';

function setup(overrides) {
  const store = initStore();
  const close = jest.fn();
  const props = { store, close, ...overrides };

  const wrapper = shallow(<CashOut {...props} />).dive();

  return {
    wrapper,
    close,
  };
}

describe('<CashOut />', () => {
  it('renders the component', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('renders error', () => {
    const { wrapper } = setup();
    wrapper.setProps({ cashOutError: 'Error happened.' });
    expect(wrapper).toMatchSnapshot();
  });

  it('renders tx', () => {
    const { wrapper, close } = setup();
    wrapper.setProps({ cashOutTx: '0x000...' });
    expect(close).toBeCalledTimes(1);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders tx without close', () => {
    const { wrapper } = setup({ close: undefined });
    wrapper.setProps({ cashOutTx: '0x000...' });
    expect(wrapper).toMatchSnapshot();
  });

  it('handles setProps without values', () => {
    const { wrapper } = setup();
    wrapper.setProps({});
    expect(wrapper).toMatchSnapshot();
  });
});
