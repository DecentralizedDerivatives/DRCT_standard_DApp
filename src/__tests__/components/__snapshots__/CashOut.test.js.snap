// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CashOut /> render() renders the component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CashOut
    cashOutError={null}
    cashOutTx=""
    getUserBalance={[Function]}
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
    storeSubscription={
      Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        },
        "unsubscribe": [Function],
      }
    }
    userBalance={0}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="modal-background"
        />,
        <div
          className="modal"
        >
          <div
            className="cashout-form"
            id="cashout-form"
          >
            <h3
              className="center-text"
            >
              Cash Out Request
            </h3>
            <h4
              className="user-balance"
            >
              Your Balance: 
              0
               Ether
            </h4>
            <Connect(ReduxForm) />
            
          </div>
        </div>,
      ],
      "className": "create-contract",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "modal-background",
          "onClick": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="cashout-form"
            id="cashout-form"
          >
            <h3
              className="center-text"
            >
              Cash Out Request
            </h3>
            <h4
              className="user-balance"
            >
              Your Balance: 
              0
               Ether
            </h4>
            <Connect(ReduxForm) />
            
          </div>,
          "className": "modal",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h3
                className="center-text"
              >
                Cash Out Request
              </h3>,
              <h4
                className="user-balance"
              >
                Your Balance: 
                0
                 Ether
              </h4>,
              <Connect(ReduxForm) />,
              "",
            ],
            "className": "cashout-form",
            "id": "cashout-form",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Cash Out Request",
                "className": "center-text",
              },
              "ref": null,
              "rendered": "Cash Out Request",
              "type": "h3",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "Your Balance: ",
                  0,
                  " Ether",
                ],
                "className": "user-balance",
              },
              "ref": null,
              "rendered": Array [
                "Your Balance: ",
                0,
                " Ether",
              ],
              "type": "h4",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "alertUserBalance": undefined,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "",
          ],
          "type": "div",
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="modal-background"
          />,
          <div
            className="modal"
          >
            <div
              className="cashout-form"
              id="cashout-form"
            >
              <h3
                className="center-text"
              >
                Cash Out Request
              </h3>
              <h4
                className="user-balance"
              >
                Your Balance: 
                0
                 Ether
              </h4>
              <Connect(ReduxForm) />
              
            </div>
          </div>,
        ],
        "className": "create-contract",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "modal-background",
            "onClick": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="cashout-form"
              id="cashout-form"
            >
              <h3
                className="center-text"
              >
                Cash Out Request
              </h3>
              <h4
                className="user-balance"
              >
                Your Balance: 
                0
                 Ether
              </h4>
              <Connect(ReduxForm) />
              
            </div>,
            "className": "modal",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <h3
                  className="center-text"
                >
                  Cash Out Request
                </h3>,
                <h4
                  className="user-balance"
                >
                  Your Balance: 
                  0
                   Ether
                </h4>,
                <Connect(ReduxForm) />,
                "",
              ],
              "className": "cashout-form",
              "id": "cashout-form",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Cash Out Request",
                  "className": "center-text",
                },
                "ref": null,
                "rendered": "Cash Out Request",
                "type": "h3",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    "Your Balance: ",
                    0,
                    " Ether",
                  ],
                  "className": "user-balance",
                },
                "ref": null,
                "rendered": Array [
                  "Your Balance: ",
                  0,
                  " Ether",
                ],
                "type": "h4",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "alertUserBalance": undefined,
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "",
            ],
            "type": "div",
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
